components/idp-mgt/org.wso2.carbon.idp.mgt/src/main/java/org/wso2/carbon/idp/mgt/IdentityProviderManager.java

```diff
@@ -1724,6 +1761,73 @@ public void addIdP(IdentityProvider identityProvider, String tenantDomain)
+            throws IdentityProviderManagementException {
+
+        // invoking the pre listeners
+        Collection<IdentityProviderMgtListener> listeners = IdPManagementServiceComponent.getIdpMgtListeners();
+        for (IdentityProviderMgtListener listener : listeners) {
+            if (listener.isEnable() && !listener.doPreAddIdP(identityProvider, tenantDomain)) {
+                return null;
+            }
+        }
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+
+        if (IdPManagementServiceComponent.getFileBasedIdPs().containsKey(identityProvider.getIdentityProviderName())
+                && !identityProvider.getIdentityProviderName().startsWith(IdPManagementConstants.SHARED_IDP_PREFIX)) {
+            //If an IDP with name starting with "SHARED_" is added from UI, It's blocked at the service class
+            // before calling this method
+            throw new IdentityProviderManagementException("Identity provider with the name" + identityProvider
+                    .getIdentityProviderName() + "exists in the file system.");
+        }
+
+        if (identityProvider.getPermissionAndRoleConfig() != null
+                && identityProvider.getPermissionAndRoleConfig().getRoleMappings() != null) {
+            verifyAndUpdateRoleConfiguration(tenantDomain, tenantId, identityProvider.getPermissionAndRoleConfig());
+        }
+
+        if (IdentityProviderManager.getInstance().getIdPByName(
+                identityProvider.getIdentityProviderName(), tenantDomain, true) != null) {
+            String msg = "An Identity Provider has already been registered with the name "
+                    + identityProvider.getIdentityProviderName() + " for tenant " + tenantDomain;
+            throw new IdentityProviderManagementException(msg);
+        }
+        StringBuilder idpName = new StringBuilder("");
+        StringBuilder metadata = new StringBuilder("");
+
+        handleMetadta(identityProvider, idpName, metadata);
+
+        validateIdPEntityId(identityProvider.getFederatedAuthenticatorConfigs(), tenantId, tenantDomain);
+        if (
+                idpName.toString().length() > 0 &&
+                        metadata.toString().length() > 0
+                ) {
+            if (SAML2SSOMetadataConverter != null) {
+                SAML2SSOMetadataConverter.saveMetadataString(tenantId, idpName.toString(), metadata.toString());
+            } else {
+                throw new  IdentityProviderManagementException("Couldn't save metadata in registry.SAML2SSOMetadataConverter is not set.");
+            }
+        }
+        ExtendedIdentityProvider idp = dao.addIdP(identityProvider, tenantId, tenantDomain);
+
+        // invoking the post listeners
+        for (IdentityProviderMgtListener listener : listeners) {
+            if (listener.isEnable() && !listener.doPostAddIdP(identityProvider, tenantDomain)) {
```

Comment:
Shouldn't we pass the created idp returned from the DAO layer to the postAddIdp method. The original identity provider passed for idp creation will not contain the generated resourceId etc.